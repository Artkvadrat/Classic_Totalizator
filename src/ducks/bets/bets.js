/* eslint-disable camelcase */
import HTTPService from '../../services/HTTPService/HTTPService';
import parseDate from '../../services/dateParse/dateParse';

export const REQUESTED_BETS = 'bets/requested';
export const RECEIVED_BETS = 'bets/received';

const requestedBets = () => ({
  type: REQUESTED_BETS
});

const receivedBets = (data) => ({
  type: RECEIVED_BETS,
  payload: data
});

export const loadBets = () => (dispatch) => {
  dispatch(requestedBets());

  return HTTPService.request({
    path: '/api/Bet/history/admin'
  }).then((data) => {
    const structured = data.betsPreviewForAdmins
      .sort((a, b) => Date.parse(b.betTime) - Date.parse(a.betTime))
      .map(
        ({
          bet_Id,
          account_Id,
          teamConfrontation,
          choice,
          eventStartime,
          betTime,
          amount,
          status
        }) => ({
          eventTitle: teamConfrontation,
          betChoice: choice,
          betTime: parseDate(betTime),
          eventTime: parseDate(eventStartime),
          betStatus: status || 'Not resolved',
          userId: account_Id,
          key: bet_Id,
          betMoney: amount
        })
      );

    dispatch(receivedBets(structured));
  });
};

const initialState = {
  isLoading: false,
  betsData: []
};

const betsReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED_BETS:
      return {
        ...state,
        isLoading: true
      };

    case RECEIVED_BETS:
      return {
        ...state,
        isLoading: false,
        betsData: action.payload
      };

    default:
      return state;
  }
};

export default betsReducer;
