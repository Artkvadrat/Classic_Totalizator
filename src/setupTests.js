import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
