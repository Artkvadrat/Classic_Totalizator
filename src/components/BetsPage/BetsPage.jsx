import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Col, Row, Skeleton, Table } from 'antd';

import useInterval from '../../services/useInterval/useInterval';
import { loadBets } from '../../ducks/bets/bets';

const BetsPage = () => {
  const dispatch = useDispatch();
  const { isLoading, betsData } = useSelector((state) => state.bets);

  useEffect(() => {
    dispatch(loadBets());
  }, [dispatch]);

  useInterval(() => {
    dispatch(loadBets());
  }, 5000);

  return isLoading ? (
    <Skeleton active />
  ) : (
    <>
      <Row justify="start" style={{ margin: 16 }}>
        <Col span={20}>
          <h1>Bets (live)</h1>
        </Col>
      </Row>

      <Table dataSource={betsData} bordered>
        <Table.ColumnGroup title="Bets">
          <Table.Column title="Time" dataIndex="betTime" key="betTime" />
          <Table.Column title="Stake" dataIndex="betChoice" key="betChoice" />
          <Table.Column title="Amount" dataIndex="betMoney" key="betMoney" />
          <Table.Column title="Status" dataIndex="betStatus" key="betStatus" />
        </Table.ColumnGroup>
        <Table.ColumnGroup title="Event">
          <Table.Column title="Match" dataIndex="eventTitle" key="eventTitle" />
          <Table.Column title="Time" dataIndex="eventTime" key="eventTime" />
        </Table.ColumnGroup>
        <Table.Column title="User" dataIndex="userId" key="userId" />
      </Table>
    </>
  );
};

export default BetsPage;
