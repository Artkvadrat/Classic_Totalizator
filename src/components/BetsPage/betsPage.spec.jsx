import React from 'react';
import { shallow, mount } from 'enzyme';
import { useDispatch, useSelector } from 'react-redux';

import { Skeleton, Table } from 'antd';

import '../../setupTests';

import { loadBets } from '../../ducks/bets/bets';
import BetsPage from './BetsPage';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn()
}));

jest.mock('../../ducks/bets/bets', () => ({
  loadBets: jest.fn()
}));

describe('BetsPage component', () => {
  const dispatch = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    loadBets.mockReturnValue(Symbol.for('bets'));

    useSelector.mockReturnValue({
      isLoading: true,
      betsData: []
    });
  });

  it('should dispatch action on mount', () => {
    expect(dispatch).not.toHaveBeenCalled();
    expect(loadBets).not.toHaveBeenCalled();

    mount(<BetsPage />);

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('bets'));
    expect(loadBets).toHaveBeenCalledTimes(1);
  });

  it('should render skeleton if loading in progress', () => {
    const result = shallow(<BetsPage />);
    const skeleton = result.find(Skeleton);

    expect(skeleton.length).toEqual(1);
    expect(skeleton.prop('active')).toEqual(true);

    expect(result).toMatchSnapshot();
  });

  describe('should render results with data', () => {
    useSelector.mockReturnValue({
      isLoading: false,
      betsData: [
        {
          eventTitle: 'Conor McGregor - France',
          betChoice: 'X',
          betTime: '12:21 27.03.2021',
          eventTime: '21:10 03.04.2021',
          betStatus: 'Active',
          userId: '3efa62a5-dfb4-4c87-bb0b-d76ef04d6026',
          key: 'ff8c2f0e-12e5-4f82-9e22-0f0362aa16ae',
          betMoney: 450
        },
        {
          eventTitle: 'Conor McGregor - France',
          betChoice: 'W1',
          betTime: '12:21 27.03.2021',
          eventTime: '21:10 03.04.2021',
          betStatus: 'Active',
          userId: '3efa62a5-dfb4-4c87-bb0b-d76ef04d6026',
          key: 'e67bc945-c328-463b-a5d5-9128c85c186d',
          betMoney: 450
        }
      ]
    });

    const result = shallow(<BetsPage />);

    it('should render table of events', () => {
      const table = result.find(Table);
      expect(table.length).toEqual(1);
    });

    expect(result).toMatchSnapshot();
  });
});
