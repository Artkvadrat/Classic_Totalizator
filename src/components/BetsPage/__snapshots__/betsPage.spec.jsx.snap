// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<Fragment>
  <Row
    justify="start"
    style={
      Object {
        "margin": 16,
      }
    }
  >
    <Col
      span={20}
    >
      <h1>
        Bets (live)
      </h1>
    </Col>
  </Row>
  <Table
    bordered={true}
    dataSource={
      Array [
        Object {
          "betChoice": "X",
          "betMoney": 450,
          "betStatus": "Active",
          "betTime": "12:21 27.03.2021",
          "eventTime": "21:10 03.04.2021",
          "eventTitle": "Conor McGregor - France",
          "key": "ff8c2f0e-12e5-4f82-9e22-0f0362aa16ae",
          "userId": "3efa62a5-dfb4-4c87-bb0b-d76ef04d6026",
        },
        Object {
          "betChoice": "W1",
          "betMoney": 450,
          "betStatus": "Active",
          "betTime": "12:21 27.03.2021",
          "eventTime": "21:10 03.04.2021",
          "eventTitle": "Conor McGregor - France",
          "key": "e67bc945-c328-463b-a5d5-9128c85c186d",
          "userId": "3efa62a5-dfb4-4c87-bb0b-d76ef04d6026",
        },
      ]
    }
    rowKey="key"
  >
    <ColumnGroup
      title="Bets"
    >
      <Column
        dataIndex="betTime"
        key="betTime"
        title="Time"
      />
      <Column
        dataIndex="betChoice"
        key="betChoice"
        title="Stake"
      />
      <Column
        dataIndex="betMoney"
        key="betMoney"
        title="Amount"
      />
      <Column
        dataIndex="betStatus"
        key="betStatus"
        title="Status"
      />
    </ColumnGroup>
    <ColumnGroup
      title="Event"
    >
      <Column
        dataIndex="eventTitle"
        key="eventTitle"
        title="Match"
      />
      <Column
        dataIndex="eventTime"
        key="eventTime"
        title="Time"
      />
    </ColumnGroup>
    <Column
      dataIndex="userId"
      key="userId"
      title="User"
    />
  </Table>
</Fragment>
`;

exports[`BetsPage component should render skeleton if loading in progress 1`] = `
<Skeleton
  active={true}
  avatar={false}
  paragraph={true}
  title={true}
/>
`;
